toolchain/docs/parse.md

```diff
@@ -0,0 +1,912 @@
+lookup is performed.
+
+The parse tree's structure corresponds to the grammar of the Carbon language. On
+valid input, there will be a 1:1 correspondence between parse tree nodes and
+tokens.
+
+A parse tree is considered _structurally valid_ if all nodes have the number of
+children that their node kind requires. On invalid input, nodes may be added
+that don't correspond to a token to maintain a structurally valid parse tree.
+When a parse tree node is marked as having an error, it will still be
+structurally valid, but its children may not match a valid grammar. Code trying
+to handle children of erroneous nodes must be prepared to handle atypical
+structures, but it may still be helpful for tools such as syntax highlighters or
+refactoring tools.
+
+In general, we favor doing the checking for whether something is allowed _in a
+particular context_ in [the check stage](check.md) instead of the parse stage,
+unless the context is very local. This is for a few reasons:
+
+-   We anticipate that the parse stage will be used to operate on invalid code
+    while still preserving as much of the intent of the author as possible, for
+    example in an IDE or a code formatter.
+-   To keep as much code out of the parse stage as possible, so it is simple and
+    fast.
+-   We are building all the infrastructure to keep track of context in the check
+    stage.
+
+These reasons explain what local context is okay: where we already have the
+contextual information at hand so there is no performance cost, and we can
+output a parse tree that still captures faithfully what the user wrote.
+Examples:
+
+-   All declaration modifiers are allowed in any order on any declaration in the
+    parse stage. Diagnosing duplicated modifiers, modifiers that conflict with
+    other modifiers, or modifiers that can't be used on a particular declaration
+    is postponed until the check stage.
+-   Rejecting a keyword after `fn` where a name is expected is done at the parse
+    stage.
+
+## Parse stack
+
+The core parser loop is `Parse::Tree::Parse`. In the loop, it pops the next
+state off the stack, and dispatches to the appropriate `Handle` function.
+
+A typical handler function pops the state first, leaving the stack ready for the
+next state. It may add nodes to the parse tree, based on the current code. If it
+needs to trigger other states, it will push them onto the stack; because it's a
+stack, the _next_ state is always pushed _last_.
+
+Operator expressions store information about current operator precedence in the
+stack as well. While this isn't necessary for most parser states, and could be
+stored separately, it's currently together because it has no impact on the size
+of a stack entry and is thus more efficient to store in one place.
+
+## Postorder tree
+
+The parse tree's storage layout is in postorder. For example, given the code:
+
+```cpp
+fn foo() -> f64 {
+  return 42;
+}
+```
+
+The node order is (with indentation to indicate nesting):
```

Comment:
The following code block is missing the indentation from the original Google Doc.

Reply:
Fixed. Looks like this was getting auto-formatted away. Other examples are fine because they're not fully valid yaml.